### PR TITLE
TW-463 Blog CQRS and ElasticSearch

### PR DESCRIPTION
**Pull request by jensfischerhh**

### CODE REVIEW

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 1]
```diff
@@ -0,0 +1,227 @@
+---
+layout: post
+title: "CQRS and Elasticsearch"
```
Comment (roseritter) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: Is this headline set in stone? 
Maybe we can find a header that is even more SEO-friendly :)
check out: http://coschedule.com/headline-analyzer

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+{% image blog/blog-cqrs-elasticsearch-database.png %}
+
+### The Read Model
+
+The storefront application renders product data for the customers visiting a merchant's shop.
+It is displaying a list of products as a result of a customer browsing a category or using the search functionality of the shop.
+When navigating to a product detail page, more product-related data is presented to the customer.
+
+Especially the need for full-text searching and faceted filtering (also called *after-search navigation*) influenced the choice to use search technology, namely Elasticsearch, as the only source for accessing product data from the storefront.
```
Comment (roseritter) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: maybe provide a link to Elasticsearch

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md (hunk 1 above)
Comment (jensfischerhh) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: sure, it's still subject to change. Let's put in some more emotion 😃

Comment (jensfischerhh) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: "Taming data synchronization with CQRS and Elasticsearch"

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+## Command and Query Responsibility Segregation
+
+Having different models and services for mutating and querying data from the same business domain can be architected using CQRS.
+
+We use a heavily normalized model for updating product data, what is called the `Command Responsibility`.
+Our service for this use case is named *product-management* and uses a relational database for persistence.
+
+The `Query Responsibility` in our scenario uses a denormalized model in a read-only fashion.
+Our service for this use case is named *product-view* and uses Elasticsearch for persistence.
```
Comment (roseritter) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: the sentence in line 60 starts the very same way ("Our service for this use case is named...")
Do you think we could start one of the sentences differently, so it doesn't sound so similar?

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+image: blog-header/searching.jpg
+categories: tech-stories
+authors: ["Jens"]
+---
+
+After having explained how we use [Multitenancy and Elasticsearch][prev-blog-post], we will continue our series of search-related blog posts by looking at a software architecture pattern called `Command and Query Responsibility Segregation` ([CQRS][fowler-cqrs]) and how we embrace it for managing and accessing product data in our ePages shops.
+
+## Different Product Data Models
+
+In our ePages software we identified two fundamentally different usage scenarios when dealing with product data, each coming with it's own data model.
```
Comment (jensfischerhh) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: 👍

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    }, {
+        "name" : "back",
+        "uri" : "https://myshop.com/images/334163-2/back.jpg"
+    }]
+}
+{% endhighlight %}
+
+The *availability* state is a computed value based on the *stock* and *threshold* numbers.
+
+This full set of attributes is only used when rendering the product detail page;
```
Comment (jensfischerhh) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: Semicolon was intentional, but could also go with a full stop.

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+## Conclusion
+
+Keeping decentrally managed data in sync is not trivial.
+Even when accepting [Eventual Constistency][fowler-eventual-consistency], you need to implement data synchronization processes and take care of storage-specific peculiarities.
+
+Asynchronous processing of events imposes another complexity by having no guarantees on the order of received messages.
+Introducing Elasticsearch [optimistic concurrency control][es-concurrency] and retrying out-of-order events could minimize this, at the cost of yet another complex implementation detail.
+
+The advantages of solving all these challenges must justify the effort involved.
+In our microservices architecture we already rely on messaging for data synchronization, so there is no additional complexity introduced.
```
Comment (jensfischerhh) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: That's not the idea. Data Sync is complex by nature, but that is complexity we already have. Separating product-mgmt from product-view does not add to the already incurred complexity.

Comment (roseritter) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: alright:)

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md (hunk 1 above)
Comment (jensfischerhh) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: "Harnessing best of DB and Elasticsearch with CQRS"

Comment (roseritter) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: @BirgitBader  what do you think?

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+title: "CQRS and Elasticsearch"
+date: "2016-12-20 09:00:00"
+image: blog-header/searching.jpg
+categories: tech-stories
+authors: ["Jens"]
+---
+
+After having explained how we use [Multitenancy and Elasticsearch][prev-blog-post], we will continue our series of search-related blog posts by looking at a software architecture pattern called `Command and Query Responsibility Segregation` ([CQRS][fowler-cqrs]) and how we embrace it for managing and accessing product data in our ePages shops.
+
+## Different Product Data Models
```
Comment (BirgitBader) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: Different product data models

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+In our ePages software we identified two fundamentally different usage scenarios when dealing with product data, each coming with its own data model.
+
+### The Write Model
+
+A merchant uses the back office application to manage all aspects of her shop, in particular it's product data.
```
Comment (BirgitBader) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: Change "back office" to "administration" or "administration area".

... all aspects of their shop...

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+This product data is stored in a relational database, where transactions and database constraints guarantee consistency and integrity.
+A simplified relational data model looks like this:
+
+{% image blog/blog-cqrs-elasticsearch-database.png %}
+
+### The Read Model
+
+The storefront application renders product data for the customers visiting a merchant's shop.
+It is displaying a list of products as a result of a customer browsing a category or using the search functionality of the shop.
```
Comment (BirgitBader) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: It displays a list of products...

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+    }]
+}
+{% endhighlight %}
+
+The *availability* state is a computed value based on the *stock* and *threshold* numbers.
+
+This full set of attributes is only used when rendering the product detail page.
+Lists of products only show a subset of these attributes, hiding *sku*, *description* and all but one *image*.
+
+## Command and Query Responsibility Segregation
```
Comment (BirgitBader) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: Command and query responsibility segregation

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+We use a heavily normalized model for updating product data, what is called the `Command Responsibility`.
+Our service for this use case is named *product-management* and uses a relational database for persistence.
+
+The `Query Responsibility` in our scenario uses a denormalized model in a read-only fashion.
+It is persisted using Elasticsearch and implemented in a service called *product-view*.
+
+By segregating these responsibilites into specialized microservices, we can use the best fitting technology available for implementing them and also scale them asymmetrically.
+This takes into account that many more customers browse the different storefronts concurrently, compared to the number of merchants updating their product data.
+
+## Data Synchronization
```
Comment (BirgitBader) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: Data synchronization

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+        "stock" : 2,
+        "threshold" : 5
+    }
+}
+{% endhighlight %}
+
+
+## Partial Updates of Documents
+
+[Updating][es-update] an already indexed complete document in Elasticsearch is pretty straight-forward:
```
Comment (BirgitBader) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: straight forward

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+    "_id" : "456",
+    "_version" : 2,
+    "found" : true,
+    "_source" : {
+        "availability" : "LOW_STOCK",
+    }
+}
+{% endhighlight %}
+
+Luckily Elasticsearch offers a feature called [partial updates][es-partial-update] to mitigate exactly this problem, leaving all remaining attributes untouched:
```
Comment (BirgitBader) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: Luckily, ...

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+Introducing Elasticsearch [optimistic concurrency control][es-concurrency] and retrying out-of-order events could minimize this, at the cost of yet another complex implementation detail.
+
+The advantages of solving all these challenges must justify the effort involved.
+In our microservices architecture we already rely on messaging for data synchronization, so there is no additional complexity introduced.
+Giving us the option to scale *product-view* individually from *product-management* is only a minor benefit, but nevertheless nice to have.
+
+By separating these two microservices, each of them can focus on its special technology stack, keeping the amount of knowledge needed to maintain each of them at minimum.
+Having all the safety of a relational database for manipulating product data as well as offering features mandatory for every e-commerce site by using Elasticsearch, is the biggest benefit.
+
+
```
Comment (BirgitBader) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: Add section:

## Related posts
include post of Ulf here as a link.

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+authors: ["Jens"]
+---
+
+After having explained how we use [Multitenancy and Elasticsearch][prev-blog-post], we will continue our series of search-related blog posts by looking at a software architecture pattern called `Command and Query Responsibility Segregation` ([CQRS][fowler-cqrs]) and how we embrace it for managing and accessing product data in our ePages shops.
+
+## Different Product Data Models
+
+In our ePages software we identified two fundamentally different usage scenarios when dealing with product data, each coming with its own data model.
+
+### The Write Model
```
Comment (jensfischerhh) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: The only established term would be "Command and Query Responsibility Segregation", so I leave that in Uppercase while changing the rest to lowercase

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md (hunk 10 above)
Comment (jensfischerhh) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: It's a known term in software architecture, hence the abbreviation CQRS. Thus I would keep it like that.

File: _posts/2016/2016-12-20-cqrs-and-elasticsearch.md (hunk 1 above)
Comment (jensfischerhh) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: _"Harnessing the best features of RDBMS and Elasticsearch with CQRS"_ gives a headline-analyzer ranking of 67, which is higher than the other alternatives. What should be the filename of the *.md file, that directly influences the blog post URL? `harnessing-the-best-features-of-rdbms-and-elasticsearch-with-cqrs.html`?

Comment (roseritter) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: maybe: rdbms-and-elasticsearch-with-cqrs.md?
 
or is that still too long?
@jensfischerhh

Comment (jensfischerhh) [inline _posts/2016/2016-12-20-cqrs-and-elasticsearch.md]: renamed the file accordingly

**Comment by jensfischerhh** [comment]
READY FOR MERGE?

**Pull request merged**